mllib/src/main/scala/org/apache/spark/ml/feature/OneHotEncoderEstimator.scala

```diff
@@ -0,0 +1,439 @@
+
+    require(inputColNames.length == outputColNames.length,
+      s"The number of input columns ${inputColNames.length} must be the same as the number of " +
+        s"output columns ${outputColNames.length}.")
+
+    require(inputColNames.length == categorySizes.length,
+      s"The number of input columns ${inputColNames.length} must be the same as the number of " +
+        s"features ${categorySizes.length} during fitting.")
+
+    val inputOutputPairs = inputColNames.zip(outputColNames)
+    val outputFields = inputOutputPairs.map { case (inputColName, outputColName) =>
+
+      require(schema(inputColName).dataType.isInstanceOf[NumericType],
+        s"Input column must be of type NumericType but got ${schema(inputColName).dataType}")
+      require(!inputFields.exists(_.name == outputColName),
+        s"Output column $outputColName already exists.")
+
+      OneHotEncoderCommon.transformOutputColumnSchema(
+        schema(inputColName), $(dropLast), outputColName)
+    }
+    verifyNumOfValues(StructType(inputFields ++ outputFields))
+  }
+
+  private def verifyNumOfValues(schema: StructType): StructType = {
+    $(outputCols).zipWithIndex.foreach { case (outputColName, idx) =>
+      val inputColName = $(inputCols)(idx)
+      val attrGroup = AttributeGroup.fromStructField(schema(outputColName))
+
+      // If the input metadata specifies number of category,
+      // compare with expected category number.
+      if (attrGroup.attributes.nonEmpty) {
+        require(attrGroup.size == categorySizes(idx), "OneHotEncoderModel expected " +
+          s"${categorySizes(idx)} categorical values for input column ${inputColName}, but " +
+            s"the input column had metadata specifying ${attrGroup.size} values.")
+      }
+    }
+    schema
+  }
+
+  @Since("2.3.0")
+  override def transform(dataset: Dataset[_]): DataFrame = {
+    if (getDropLast && getHandleInvalid == OneHotEncoderEstimator.SKIP_INVALID) {
+      throw new IllegalArgumentException("When Param handleInvalid is set to " +
+        s"${OneHotEncoderEstimator.SKIP_INVALID}, Param dropLast can't be true, " +
+        "because last category and invalid values will conflict in encoded vector.")
+    }
+
+    val transformedSchema = transformSchema(dataset.schema, logging = true)
+
+    val encodedColumns = encoders.zipWithIndex.map { case (encoder, idx) =>
+      val inputColName = $(inputCols)(idx)
+      val outputColName = $(outputCols)(idx)
+
+      val outputAttrGroupFromSchema =
+        AttributeGroup.fromStructField(transformedSchema(outputColName))
+
+      val metadata = if (outputAttrGroupFromSchema.size < 0) {
+        OneHotEncoderCommon.createAttrGroupForAttrNames(outputColName, false,
+          categorySizes(idx)).toMetadata()
+      } else {
+        outputAttrGroupFromSchema.toMetadata()
+      }
+
+      encoder(col(inputColName).cast(DoubleType)).as(outputColName, metadata)
+    }
+    val allCols = Seq(col("*")) ++ encodedColumns
```

Comment:
Just looking at this in the context of the previous multi-col work done (`Bucketizer` and `Imputer`) is there any real performance difference between this and the use of `dataset.withColumns(...)`?

Reply:
Before running real performance test, I guess the difference might not be so significant.

Besides performance, this multi-column support on `OneHotEncoder` can also benefit user convenience when using its API. For example scikit-learn's OneHotEncoder is also with multi-column support.

Reply:
If we can get the number of values for multi-column in https://github.com/apache/spark/pull/19527/files#r145457081, the performance of fitting can be improved.

Reply:
Performance test results are attached now. Existing `OneHotEncoder` only performs `transform`. So if we consider the time of fitting and transforming in this estimator/model with existing one hot encoder, this multi-column approach is faster.